File: src/main/java/com/leijendary/spring/notificationtemplate/repository/DeviceRepository.java
package com.leijendary.spring.notificationtemplate.repository;

import com.leijendary.spring.notificationtemplate.model.Device;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;

import java.util.Optional;

public interface DeviceRepository extends JpaRepository<Device, Long> {

    Optional<Device> findFirstByDeviceId(final String deviceId);

    @Modifying
    void deleteByDeviceId(final String deviceId);
}

File: src/main/kotlin/com/leijendary/spring/template/notification/repository/DeviceRepository.kt
package com.leijendary.spring.template.notification.repository

import com.leijendary.spring.template.notification.core.exception.ResourceNotFoundException
import com.leijendary.spring.template.notification.entity.Device
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.transaction.annotation.Transactional
import java.util.*
import java.util.stream.Stream

private val sourceToken = listOf("data", "Device", "token")

interface DeviceRepository : JpaRepository<Device, Long> {
    @Transactional(readOnly = true)
    fun findFirstByToken(token: String): Device?

    @Transactional(readOnly = true)
    fun findFirstByUserIdAndToken(userId: UUID, token: String): Device?

    @Transactional(readOnly = true)
    fun findFirstByUserIdAndTokenOrThrow(userId: UUID, token: String): Device {
        return findFirstByUserIdAndToken(userId, token) ?: throw ResourceNotFoundException(sourceToken, token)
    }

    @Transactional(readOnly = true)
    fun streamByUserId(userId: UUID): Stream<Device>
}
